### PR TITLE
Introduce JSON_EXTRACT function

### PR DESCRIPTION
**Pull request by wey-gu**
close: #3513

Note, we don't support the path argument in this phase

## What type of PR is this?
- [ ] bug
- [x] feature
- [ ] enhancement

#### Description:

It takes a string and returns a [map](https://docs.nebula-graph.io/3.2.0/3.ngql-guide/3.data-types/8.map/):

```
YIELD JSON_EXTRACT('{"a": 1, "b": 2}') AS map_foo
+-------------+
| map_foo     |
+-------------+
| {a: 1, b: 2}|
+-------------+
```

**highlight** ( should be included in docs, too)
- "foo", "1", "true" are all valid JSON String though(RFC8259), they are treated as BAD_DATA here due to this function is trying to return either empty/null or Map() in its return value, thus "foo" will be considered to be dirty value.

```
  Scenario: Test Cases With Invalid JSON String
    When executing query:
      """
      YIELD JSON_EXTRACT('fuzz') AS result;
      """
    Then the result should be, in any order:
      | result   |
      | BAD_DATA |
```

**limitation**
- Only Bool, Double, Int, String value and null are supported
- Only depth-1 nested Map is supported now.

Here nested Map depth is 1, it's parsed perfectly:

```sql
> YIELD JSON_EXTRACT('{"a": 1, "b": 0.2, "c": {"d": true}}')
+------------------------------------------------------+
| JSON_EXTRACT("{"a": 1, "b": 0.2, "c": {"d": true}}") |
+------------------------------------------------------+
| {a: 1, b: 0.2, c: {d: true}}                         |
+------------------------------------------------------+
```

Here nested Map depth is 2, the nested item is left as an empty Map():

```sql
> YIELD JSON_EXTRACT('{"a": "foo", "b": 0.2, "c": {"d": {"e": 0.1}}}')
+----------------------------------------------------------------+
| JSON_EXTRACT("{"a": "foo", "b": 0.2, "c": {"d": {"e": 0.1}}}") |
+----------------------------------------------------------------+
| {a: "foo", b: 0.2, c: {}}                                      |
+----------------------------------------------------------------+
```

## Checklist:
Tests:
- [x] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [x] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:

Added function: `JSON_EXTRACT` to enable parsing JSON format string into Map.

### CODE REVIEW

File: src/common/datatypes/Map.h [hunk 1]
```diff
@@ -22,6 +24,22 @@ struct Map {
...
+  // constructor to convert from json string
+  explicit Map(const std::string& json) {
+    auto obj = folly::parseJson(json);
+    for (auto& kv : obj.items()) {
+      kvs.emplace(kv.first.getString(), Value(kv.second));
+    }
+  }
+
+  // constructor to convert from Value::Type::STRING in JSON format
+  explicit Map(const Value& value) {
```
Comment (jievince) [inline src/common/datatypes/Map.h:36]: Add a Dcheck to check the value type is str, or just delete this interface.

Comment (wey-gu) [reply src/common/datatypes/Map.h:36]: Got it, done by removing it :)

**Review by jievince — APPROVED** [review]
LGTM

File: src/common/datatypes/Map.h [hunk 2]
```diff
@@ -28,7 +29,23 @@ struct Map {
...
+      } else if (kv.second.isBool()) {
+        kvs.emplace(kv.first.asString(), Value(kv.second.asBool()));
+      } else if (kv.second.isNull()) {
+        kvs.emplace(kv.first.asString(), Value());
+      } else {
+        LOG(ERROR) << "Only string, int64, double, bool, null are supported in "
+                      "parseJson to construct Map, now trying to parse from: "
+                   << kv.second.typeName();
+      }
     }
```
Comment (jievince) [inline src/common/datatypes/Map.h:49]: It's better to move the implementation to .cpp file.

**Review by jievince — COMMENTED** [review]
The ctest failed, please fix it.

File: src/common/datatypes/Map.h (hunk 2 above)
Comment (wey-gu) [reply src/common/datatypes/Map.h:49]: > The ctest failed, please fix it.

working on it now,

> It's better to move the implementation to .cpp file.

will do so.

Comment (wey-gu) [reply src/common/datatypes/Map.h]: done :)

**Review by jievince — APPROVED** [review]
Lgtm

**Review by jievince — DISMISSED** [review]
Lgtm

**Comment by Shylock-Hg** [comment]
Plz add test in tck.

**Comment by wey-gu** [comment]
@Shylock-Hg @jievince @Sophie-Xie it's now ready and rebased :)
Thanks!

**Review by jievince — APPROVED** [review]

File: src/common/function/FunctionManager.cpp [hunk 3]
```diff
@@ -2766,6 +2771,34 @@ FunctionManager::FunctionManager() {
...
+    attr.body_ = [](const auto &args) -> Value {
+      if (!args[0].get().isStr()) {
+        return Value::kNullBadType;
+      }
+      auto json = args[0].get().getStr();
+
+      // invalid string to json will be caught and returned as null
+      try {
+        auto obj = folly::parseJson(json);
+        if (!obj.isObject()) {
```
Comment (Shylock-Hg) [inline src/common/function/FunctionManager.cpp:2789]: Why check this?

Comment (wey-gu) [reply src/common/function/FunctionManager.cpp:2789]: Or invalid json string will end up abnormal here, led the connection hanging.

Comment (Shylock-Hg) [reply src/common/function/FunctionManager.cpp:2789]: `"string"` is not object but also valid json.

Comment (wey-gu) [reply src/common/function/FunctionManager.cpp:2789]: Oh, thanks @Shylock-Hg , now I know "foo" is indeed a valid JSON according to RFC 8259, I will revisit here and fix this case!

Comment (wey-gu) [reply src/common/function/FunctionManager.cpp:2789]: @Shylock-Hg , in SQL DB, "foo" was parsed as NULL, thus I think checking isObject() is still reasonable here, what do you think?

Screen Shot 2022-10-24 at 10 19 17

Comment (wey-gu) [reply src/common/function/FunctionManager.cpp:2789]: aligned with @Shylock-Hg to go with option-1, I have updated the highlight in this PR to ensure this behavior will be emphasized in docs.

**Review by Shylock-Hg — APPROVED** [review]

**Comment by foesa-yang** [comment]
https://github.com/vesoft-inc/nebula-docs-cn/pull/2268

**Pull request merged**